hop/subscribe.py

```diff
@@ -6,28 +6,69 @@
 
 import argparse
 import json
+import warnings
 
 from . import cli
 from .io import Stream
-from .models import GCNCircular
+from .models import GCNCircular, VOEvent
 
 
-def print_gcn(gcn_dict, json_dump=False):
-    """Parse a gcn dictionary and print to stdout.
+def classify_msg(msg):
+    """Check and classify the format of a message obtained from an ADC
+    stream, use it to instantiate a data model corresponding to that format.
 
     Args:
-      gcn_dict:  the dictionary object containing a GCN-formatted message
+      msg:       raw message from an ADC stream
 
     Returns:
-      None
+      gcn:       dataclass model object for the raw message
 
     """
+    # check for msg format using standard-specific flags
+    voevent_flag = "ivorn"
+    gcncir_flag = "GCN CIRCULAR"
+
+    # VOEvent:
+    if voevent_flag in msg:
+        gcn = VOEvent(**msg)
+        status_str = "## Parsing a VOEvent"
+    # GCN circular:
+    elif gcncir_flag in msg:
+        gcn = GCNCircular(**msg)
+        status_str = "## Parsing a GCN Circular"
+    else:
+        try:
+            gcn = GCNCircular(**msg)
+            status_str = "## Parsing a hop-published message"
+        except:
+            warnings.warn("## Warning: message format not recognized; dumping as json")
+            gcn = msg
+            status_str = "## Dumping unknown message as json"
+
+            print(status_str)
+            print_gcn(gcn, json_dump=True)
+            return None
```

Comment:
What is the rationale behind the second try/except? I'd imagine is it isn't recognized as falling into a VOEvent or GCNCircular above, then it should just be treated as a message whose type that can't be recognized automatically, but maybe I'm missing something.

I'm not sure how I feel about the warning, but I don't think we discussed whether the hop client should or should not handle arbitrary message formats.

Reply:
I added the `except` to handle unknown input (like a string instead of a file) during testing, though I think `hop` should fail in that case rather than attempt printing. I'll remove this. Later we should plan to figure out whether hop-client should handle arbitrary input (along with message typing itself).